src/transformers/models/big_bird/modeling_big_bird.py

```diff
@@ -0,0 +1,2695 @@
+
+        second_last_key_mat = torch.cat(
+            [
+                blocked_key_matrix[:, :, 0],
+                blocked_key_matrix[:, :, -3],
+                blocked_key_matrix[:, :, -2],
+                blocked_key_matrix[:, :, -1],
+                gathered_key[:, :, -1],
+            ],
+            dim=2,
+        )  # [b, h, (4+r)*wn, -1]
+        second_last_value_mat = torch.cat(
+            [
+                blocked_value_matrix[:, :, 0],
+                blocked_value_matrix[:, :, -3],
+                blocked_value_matrix[:, :, -2],
+                blocked_value_matrix[:, :, -1],
+                gathered_value[:, :, -1],
+            ],
+            dim=2,
+        )  # [b, h, (4+r)*wn, -1]
+
+        # [b, h, wm, -1] x [b, h, (4+r)*wn, -1] ==> [b, h, wm, (4+r)*wn]
+        second_last_product = self.torch_bmm_nd_transpose(blocked_query_matrix[:, :, -2], second_last_key_mat, ndim=4)
+        second_last_seq_pad = torch.cat(
+            [
+                to_mask[:, :, :, :wn],
+                to_mask[:, :, :, -3 * wn :],
+                torch.ones([b, 1, 1, r * wn], device=context_layer.device, dtype=context_layer.dtype),
+            ],
+            dim=3,
+        )
+        second_last_rand_pad = torch.cat(
+            [
+                torch.ones([b, h, wm, 4 * wn], device=context_layer.device, dtype=context_layer.dtype),
+                rand_mask[:, :, -1],
+            ],
+            dim=3,
+        )
+        second_last_product = second_last_product * rsqrt_d
+        second_last_product += (1.0 - torch.minimum(second_last_seq_pad, second_last_rand_pad)) * -10000.0
+        second_last_attn_weights = F.softmax(second_last_product, dim=-1)  # [b, h, wm, (4+r)*wn]
+
+        # [b, h, wm, (4+r)*wn] x [b, h, (4+r)*wn, -1] ==> [b, h, wm, -1]
+        second_last_context_layer = self.torch_bmm_nd(second_last_attn_weights, second_last_value_mat, ndim=4)
+        second_last_context_layer.unsqueeze_(2)
+
+        # last block is global q[-1] x (k[0], k[1], k[2], k[3], .... )
+
+        # [b, h, wm, -1] x [b, h, n, -1] ==> [b, h, wm, n]
+        last_product = self.torch_bmm_nd_transpose(blocked_query_matrix[:, :, -1], key_layer, ndim=4)
+        last_product = last_product * rsqrt_d
+        last_product += (1.0 - to_mask) * -10000.0
+        last_attn_weights = F.softmax(last_product, dim=-1)  # [b, h, wm, n]
+
+        # [b, h, wm, n] x [b, h, n, -1] ==> [b, h, wm, -1]
+        last_context_layer = self.torch_bmm_nd(last_attn_weights, value_layer, ndim=4)
+        last_context_layer.unsqueeze_(2)
+        context_layer = torch.cat(
+            [first_context_layer, second_context_layer, context_layer, second_last_context_layer, last_context_layer],
+            dim=2,
+        )
+        context_layer = context_layer.view((b, h, m, -1)) * from_mask
+        context_layer = torch.transpose(context_layer, 1, 2)
+
+        if output_attentions:
+            # TODO(PVP): need to verify if below code is correct
```

Comment:
Let's leave it like this for now! -> I'll try to check-out after merging whats going on here exactly